userscript-preferences.js

```diff
@@ -0,0 +1,230 @@
+/*
+
+All callers are required to have loaded jQuery themselves.
+
+Function `add_userscript_settings` requires 2 arguments.
+
+The first a text-string to be used as a section heading; something a user
+will recognize associates these settings with your userscript.
+
+The second an array of option objects.
+Each option object corresponds to a single preference.
+The keys 'type', 'label', 'value' and 'options' are used to render the
+settings form field.
+The key 'callback' is optional. The provided function is called with a single
+value - the value entered / selected by the user.
+The key 'localStorage' is optional. This will be used as a key in the
+browser's localStorage. If provided, any 'value' argument will be ignored
+
+Example user-code:
+
+if ( window.location.pathname === "/preferences" ) {
+    add_userscript_settings(
+        "My Heading",
+        [
+            {
+                label: "Text area",
+                type: "text",
+                value: "value 1",
+                callback: function(value) {
+                    alert(value);
+                }
+            },
+            {
+                label: "Checkbox",
+                type: "checkbox",
+                value: false,
+                callback: function(value) {
+                    alert(value);
+                }
+            },
+            {
+                label: "Radio buttons",
+                type: "radio",
+                value: "one",
+                options: ["one", "two"],
+                callback: function(value) {
+                    alert(value);
+                }
+            }
+        ]
+    );
+}
+
+*/
+
+var core_prefs;
+
+function add_userscript_settings(def) {
+    "use strict";
+
+    if ( "/preferences" !== window.location.pathname ) {
+        return;
+    }
+
+    let stored_defaults = {};
+    let stored_defaults_str = localStorage.getItem( def.key );
+    if ( stored_defaults_str ) {
+        try {
+            stored_defaults = JSON.parse( stored_defaults_str );
+        }
+        catch (error) {
+            // console.log("caught error");
+            // console.log(error);
+        }
+    }
+    let def_defaults = {};
+    if ( def.hasOwnProperty("defaults") ) {
+        def_defaults = def.defaults;
+    }
+
+
+    if ( !core_prefs ) {
+        core_prefs = $("#preferences");
+    }
+
+    let label = def.label || def.key;
+    core_prefs.append( `<h2>UserScript: ${label}</h2>` );
```

Comment:
BTW, if you use `<h2 class="settings-header">` , you get a horizontal separator bar above it for free (via CSS that sets `border-top`). That can help further visually distinguish the userscript settings region from Tau settings (and from other userscripts' settings regions).

Reply:
Thanks - I've added that